File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) {
            return null;
        }
        Queue<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int size = q.size();
            // System.out.println(size);
            Node currRoot = q.poll();
            // System.out.println(currRoot.val);

            if (currRoot.left != null) {
                q.add(currRoot.left);
            }
            if (currRoot.right != null) {
                q.add(currRoot.right);
            }
            for (int i = 1; i < size; i++) {
                Node temp = q.poll();

                currRoot.next = temp;
                currRoot = temp;
                if (currRoot.left != null) {
                    q.add(currRoot.left);
                }
                if (currRoot.right != null) {
                    q.add(currRoot.right);
                }
                // temp = currRoot;
            }
        }
        return root;
    }
}

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if root is None:
            return None

        queue = deque()
        queue.append(root)

        while queue:
            size = len(queue)
            currRoot = queue.popleft()

            if(currRoot.left):
                queue.append(currRoot.left)
            
            if currRoot.right:
                queue.append(currRoot.right)
            
            for i in range(1, size):
                temp = queue.popleft()

                currRoot.next = temp
                currRoot = temp

                if currRoot.left:
                    queue.append(currRoot.left)
                
                if currRoot.right:
                    queue.append(currRoot.right)
        
        return root
    
#BFS
#TC: O(n)
#SC: O(n)